File: Week-2/Task-4/src/main/java/StudentGradeProcessor.java
import java.util.*;
import java.util.stream.Collectors;

/**
 * Data class representing a student with their grades
 */
class Student {
    private String name;
    private List<Double> grades;
    
    public Student(String name, List<Double> grades) {
        this.name = name;
        this.grades = grades != null ? new ArrayList<>(grades) : new ArrayList<>();
    }
    
    public String getName() { return name; }
    public List<Double> getGrades() { return new ArrayList<>(grades); }
    
    @Override
    public String toString() {
        return "Student{name='" + name + "', grades=" + grades + "}";
    }
}

/**
 * Data class representing a problem student with their issue
 */
class ProblemStudent {
    private String name;
    private String issue;
    
    public ProblemStudent(String name, String issue) {
        this.name = name;
        this.issue = issue;
    }
    
    public String getName() { return name; }
    public String getIssue() { return issue; }
    
    @Override
    public String toString() {
        return "ProblemStudent{name='" + name + "', issue='" + issue + "'}";
    }
}

/**
 * Data class containing the processing results
 */
class ProcessingResults {
    private int totalStudents;
    private int passingStudents;
    private int failingStudents;
    private double averageGrade;
    private Map<String, Integer> gradeDistribution;
    private List<String> recommendations;
    private List<ProblemStudent> problemStudents;
    
    public ProcessingResults() {
        this.totalStudents = 0;
        this.passingStudents = 0;
        this.failingStudents = 0;
        this.averageGrade = 0.0;
        this.gradeDistribution = new HashMap<>();
        this.recommendations = new ArrayList<>();
        this.problemStudents = new ArrayList<>();
    }
    
    // Getters
    public int getTotalStudents() { return totalStudents; }
    public int getPassingStudents() { return passingStudents; }
    public int getFailingStudents() { return failingStudents; }
    public double getAverageGrade() { return averageGrade; }
    public Map<String, Integer> getGradeDistribution() { return new HashMap<>(gradeDistribution); }
    public List<String> getRecommendations() { return new ArrayList<>(recommendations); }
    public List<ProblemStudent> getProblemStudents() { return new ArrayList<>(problemStudents); }
    
    // Setters
    public void setTotalStudents(int totalStudents) { this.totalStudents = totalStudents; }
    public void setPassingStudents(int passingStudents) { this.passingStudents = passingStudents; }
    public void setFailingStudents(int failingStudents) { this.failingStudents = failingStudents; }
    public void setAverageGrade(double averageGrade) { this.averageGrade = averageGrade; }
    
    // Helper methods
    public void addRecommendation(String recommendation) {
        this.recommendations.add(recommendation);
    }
    
    public void addProblemStudent(ProblemStudent problemStudent) {
        this.problemStudents.add(problemStudent);
    }
    
    public void incrementGradeDistribution(String gradeRange) {
        this.gradeDistribution.merge(gradeRange, 1, Integer::sum);
    }
    
    @Override
    public String toString() {
        return "ProcessingResults{" +
                "totalStudents=" + totalStudents +
                ", passingStudents=" + passingStudents +
                ", failingStudents=" + failingStudents +
                ", averageGrade=" + averageGrade +
                ", gradeDistribution=" + gradeDistribution +
                ", recommendations=" + recommendations +
                ", problemStudents=" + problemStudents +
                '}';
    }
}

/**
 * Custom exception for invalid grade threshold
 */
class InvalidGradeThresholdException extends IllegalArgumentException {
    public InvalidGradeThresholdException(String message) {
        super(message);
    }
}

/**
 * Custom exception for empty student data
 */
class EmptyStudentDataException extends IllegalArgumentException {
    public EmptyStudentDataException(String message) {
        super(message);
    }
}

/**
 * Main class for processing student grade data and generating performance analytics.
 * This is the Java equivalent of the Python process_student_grades function.
 */
public class StudentGradeProcessor {
    
    /**
     * Process student grade data and generate performance analytics.
     * 
     * @param studentData List of Student objects containing student information
     * @param gradeThreshold Minimum passing grade (default: 70)
     * @param maxAttempts Maximum number of attempts allowed (default: 3)
     * @return ProcessingResults object with statistics and recommendations
     * @throws EmptyStudentDataException if student data is null or empty
     * @throws InvalidGradeThresholdException if grade threshold is invalid
     * @throws RuntimeException if any other error occurs during processing
     */
    public static ProcessingResults processStudentGrades(
            List<Student> studentData, 
            int gradeThreshold, 
            int maxAttempts) {
        
        // Input validation
        if (studentData == null || studentData.isEmpty()) {
            throw new EmptyStudentDataException("Student data cannot be empty");
        }
        
        if (gradeThreshold < 0 || gradeThreshold > 100) {
            throw new InvalidGradeThresholdException("Grade threshold must be between 0 and 100");
        }
        
        ProcessingResults results = new ProcessingResults();
        double totalGradeSum = 0.0;
        int validStudents = 0;
        
        try {
            for (Student student : studentData) {
                // Validate student data structure
                if (student == null || student.getName() == null || student.getGrades() == null) {
                    continue;
                }
                
                String studentName = student.getName();
                List<Double> grades = student.getGrades();
                
                // Validate grades data
                if (grades.isEmpty()) {
                    results.addProblemStudent(new ProblemStudent(studentName, "No valid grades found"));
                    continue;
                }
                
                // Calculate student statistics
                List<Double> validGrades = grades.stream()
                    .filter(grade -> grade != null && grade >= 0.0 && grade <= 100.0)
                    .collect(Collectors.toList());
                
                if (validGrades.isEmpty()) {
                    results.addProblemStudent(new ProblemStudent(studentName, "No valid grades found"));
                    continue;
                }
                
                // Calculate average grade for this student
                double studentAverage = validGrades.stream()
                    .mapToDouble(Double::doubleValue)
                    .average()
                    .orElse(0.0);
                
                totalGradeSum += studentAverage;
                validStudents++;
                
                // Update grade distribution - match Python format exactly
                int gradeRangeStart = (int) (studentAverage / 10) * 10;
                String gradeRange = String.format("%.1f-%.1f", (double)gradeRangeStart, (double)(gradeRangeStart + 9));
                results.incrementGradeDistribution(gradeRange);
                
                // Determine pass/fail status
                if (studentAverage >= gradeThreshold) {
                    results.setPassingStudents(results.getPassingStudents() + 1);
                } else {
                    results.setFailingStudents(results.getFailingStudents() + 1);
                }
                
                // Check for multiple attempts (regardless of pass/fail status)
                if (validGrades.size() > maxAttempts) {
                    results.addRecommendation(
                        String.format("Student %s has %d attempts. Consider academic intervention.", 
                            studentName, validGrades.size())
                    );
                }
            }
            
            // Calculate overall statistics
            results.setTotalStudents(validStudents);
            
            if (validStudents > 0) {
                double averageGrade = totalGradeSum / validStudents;
                // Round to 2 decimal places (equivalent to Python's round())
                results.setAverageGrade(Math.round(averageGrade * 100.0) / 100.0);
            } else {
                results.setAverageGrade(0.0);
            }
            
            return results;
            
        } catch (Exception e) {
            throw new RuntimeException("Error processing student data: " + e.getMessage(), e);
        }
    }
    
    /**
     * Overloaded method with default parameters (Java doesn't support default parameters like Python)
     */
    public static ProcessingResults processStudentGrades(List<Student> studentData) {
        return processStudentGrades(studentData, 70, 3);
    }
    
    /**
     * Overloaded method with grade threshold only
     */
    public static ProcessingResults processStudentGrades(List<Student> studentData, int gradeThreshold) {
        return processStudentGrades(studentData, gradeThreshold, 3);
    }
    
    /**
     * Main method for testing and demonstration
     */
    public static void main(String[] args) {
        // Test data
        List<Student> testStudents = Arrays.asList(
            new Student("Alice Johnson", Arrays.asList(85.0, 92.0, 78.0, 88.0)),
            new Student("Bob Smith", Arrays.asList(65.0, 72.0, 68.0, 71.0)),
            new Student("Carol Davis", Arrays.asList(95.0, 98.0, 92.0, 96.0)),
            new Student("David Wilson", Arrays.asList(45.0, 52.0, 48.0, 55.0)),
            new Student("Eve Brown", Arrays.asList(75.0, 82.0, 79.0, 85.0))
        );
        
        try {
            ProcessingResults results = processStudentGrades(testStudents, 70, 3);
            System.out.println("Processing Results:");
            System.out.println("Total Students: " + results.getTotalStudents());
            System.out.println("Passing Students: " + results.getPassingStudents());
            System.out.println("Failing Students: " + results.getFailingStudents());
            System.out.println("Average Grade: " + results.getAverageGrade());
            System.out.println("Grade Distribution: " + results.getGradeDistribution());
            System.out.println("Recommendations: " + results.getRecommendations());
            System.out.println("Problem Students: " + results.getProblemStudents());
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
} 

File: Week-2/Task-4/student_grade_processor.py
def process_student_grades(student_data, grade_threshold=70, max_attempts=3):
    """
    Process student grade data and generate performance analytics.
    
    Args:
        student_data (list): List of dictionaries containing student information
        grade_threshold (int): Minimum passing grade (default: 70)
        max_attempts (int): Maximum number of attempts allowed (default: 3)
    
    Returns:
        dict: Processed results with statistics and recommendations
    """
    if not student_data:
        raise ValueError("Student data cannot be empty")
    
    if grade_threshold < 0 or grade_threshold > 100:
        raise ValueError("Grade threshold must be between 0 and 100")
    
    results = {
        'total_students': 0,
        'passing_students': 0,
        'failing_students': 0,
        'average_grade': 0.0,
        'grade_distribution': {},
        'recommendations': [],
        'problem_students': []
    }
    
    total_grade_sum = 0
    valid_students = 0
    
    try:
        for student in student_data:
            # Validate student data structure
            if not isinstance(student, dict):
                continue
                
            if 'name' not in student or 'grades' not in student:
                continue
            
            student_name = student['name']
            grades = student['grades']
            
            # Validate grades data
            if not isinstance(grades, list) or len(grades) == 0:
                results['problem_students'].append({
                    'name': student_name,
                    'issue': 'No valid grades found'
                })
                continue
            
            # Calculate student statistics
            valid_grades = []
            for grade in grades:
                if isinstance(grade, (int, float)) and 0 <= grade <= 100:
                    valid_grades.append(grade)
            
            if len(valid_grades) == 0:
                results['problem_students'].append({
                    'name': student_name,
                    'issue': 'No valid grades found'
                })
                continue
            
            # Calculate average grade for this student
            student_average = sum(valid_grades) / len(valid_grades)
            total_grade_sum += student_average
            valid_students += 1
            
            # Update grade distribution
            grade_range = f"{(student_average // 10) * 10}-{(student_average // 10) * 10 + 9}"
            results['grade_distribution'][grade_range] = results['grade_distribution'].get(grade_range, 0) + 1
            
            # Determine pass/fail status
            if student_average >= grade_threshold:
                results['passing_students'] += 1
            else:
                results['failing_students'] += 1
                
                # Check for multiple attempts
                if len(valid_grades) > max_attempts:
                    results['recommendations'].append(
                        f"Student {student_name} has {len(valid_grades)} attempts. "
                        f"Consider academic intervention."
                    )
        
        # Calculate overall statistics
        results['total_students'] = valid_students
        
        if valid_students > 0:
            results['average_grade'] = round(total_grade_sum / valid_students, 2)
        else:
            results['average_grade'] = 0.0
        
        # Generate additional recommendations
        if results['failing_students'] > results['passing_students']:
            results['recommendations'].append(
                "More than 50% of students are failing. Consider curriculum review."
            )
        
        if results['average_grade'] < grade_threshold:
            results['recommendations'].append(
                f"Class average ({results['average_grade']}) is below threshold ({grade_threshold}). "
                f"Consider additional support."
            )
        
        return results
        
    except Exception as e:
        raise RuntimeError(f"Error processing student data: {str(e)}")


# Example usage and test data
if __name__ == "__main__":
    # Test data
    test_students = [
        {
            'name': 'Alice Johnson',
            'grades': [85, 92, 78, 88]
        },
        {
            'name': 'Bob Smith',
            'grades': [65, 72, 68, 71]
        },
        {
            'name': 'Carol Davis',
            'grades': [95, 98, 92, 96]
        },
        {
            'name': 'David Wilson',
            'grades': [45, 52, 48, 55]
        },
        {
            'name': 'Eve Brown',
            'grades': [75, 82, 79, 85]
        }
    ]
    
    try:
        results = process_student_grades(test_students, grade_threshold=70, max_attempts=3)
        print("Processing Results:")
        print(f"Total Students: {results['total_students']}")
        print(f"Passing Students: {results['passing_students']}")
        print(f"Failing Students: {results['failing_students']}")
        print(f"Average Grade: {results['average_grade']}")
        print(f"Grade Distribution: {results['grade_distribution']}")
        print(f"Recommendations: {results['recommendations']}")
        print(f"Problem Students: {results['problem_students']}")
    except Exception as e:
        print(f"Error: {e}") 